src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/MemoryBuilder`1.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Razor;
+
+/// <summary>
+///  Temporary builder that uses <see cref="ArrayPool{T}"/> to back a <see cref="Memory{T}"/>.
+/// </summary>
+internal ref struct MemoryBuilder<T>
+{
+    private Memory<T> _memory;
+    private T[]? _arrayFromPool;
+    private int _index;
+
+    public MemoryBuilder(int initialCapacity = 1)
```

Comment:
Good point. And it might be even better to make it 0 by default, to avoid hitting the ArrayPool until it's actually used. That'd make `new MemoryBuilder<T>` and `default(MemoryBuilder<T>)` equivalent, which seems proper.